File: matrix_multiplication/matrix_multiplication.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

#define N 1000
#define M 1000

int main(int argc, char** argv)
{
    int rank, size, i, j, k;
    double A[N][M], B[M][N], C[N][N];
    double startTime, endTime;

    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    int chunk_size = N / size;
    int start = rank * chunk_size;
    int end = (rank + 1) * chunk_size;

    for (i = 0; i < N; i++) {
        for (j = 0; j < M; j++) {
            A[i][j] = (double)rand() / RAND_MAX;
        }
    }

    for (i = 0; i < M; i++) {
        for (j = 0; j < N; j++) {
            B[i][j] = (double)rand() / RAND_MAX;
        }
    }

    startTime = MPI_Wtime();

    for (i = start; i < end; i++) {
        for (j = 0; j < N; j++) {
            C[i][j] = 0.0;
            for (k = 0; k < M; k++) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }

    MPI_Allreduce(MPI_IN_PLACE, C, N*N, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);

    endTime = MPI_Wtime();

    if (rank == 0) {
        printf("Matrix multiplication took %f seconds\n", endTime - startTime);
    }

    MPI_Finalize();
    return 0;
}


// in c we need to compile first:
//mpicc -o matrix_multiplication matrix_multiplication.c

//then run:
//mpirun -np 8 ./matrix_multiplication

File: matrix_multiplication/matrix_multiplication.py
from mpi4py import MPI
import numpy as np

# Initialize MPI
comm = MPI.COMM_WORLD
size = comm.Get_size()
rank = comm.Get_rank()

# Define the size of the matrices
N = 1000
M = 1000

# Divide the work among the processes
chunk_size = N // size
start = rank * chunk_size
end = (rank + 1) * chunk_size

# Create the matrices
A = np.random.rand(N, M)
B = np.random.rand(M, N)
C = np.zeros((N, N))

# Perform the matrix multiplication in parallel
for i in range(start, end):
    for j in range(N):
        for k in range(M):
            C[i][j] += A[i][k] * B[k][j]

# Combine the results from all processes
comm.Allreduce(MPI.IN_PLACE, C, op=MPI.SUM)

# Print the result
if rank == 0:
    print(C)

"""
Since this program is written in Python and uses the MPI4Py library, 
there is no need to compile the code before running
"""
#run:
#mpirun -np 8 python matrix_multiplication.py

